engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala

```diff
@@ -780,30 +779,47 @@ object WorkflowExecutionActor {
 
   object ExpressionKey {
     def apply(declaration: ExpressionNode, index: ExecutionIndex): ExpressionKey = {
-      DynamicDeclarationKey(declaration, index)
+      IntermediateValueKey(declaration, index)
+    }
+    def apply(declaration: ExpressionBasedGraphOutputNode): ExpressionKey = {
+      ExpressionBasedWorkflowOutputKey(declaration)
     }
   }
 
   sealed trait ExpressionKey extends JobKey {
-    override val node: ExpressionNode
-    override val attempt = 1
-    override val tag = s"Expression-${node.unqualifiedName}:${index.fromIndex}:$attempt"
-  }
-
-  case class DynamicDeclarationKey(node: ExpressionNode, index: ExecutionIndex) extends ExpressionKey {
     import lenthall.validation.ErrorOr._
     import lenthall.validation.Validation._
 
-    lazy val inputs: Map[String, InputPort] = node.instantiatedExpression.inputMapping
+    override val attempt = 1
+    override lazy val tag = s"Expression-${node.localName}:${index.fromIndex}:$attempt"
+    
+    protected def instantiatedExpression: InstantiatedExpression
+    def womType: WdlType
+    def singleOutputPort: OutputPort
+    
+    private lazy val inputs: Map[String, InputPort] = instantiatedExpression.inputMapping
     lazy val upstreamPorts: Map[String, OutputPort] = inputs map {
       case (key, input) => key -> input.upstream
     }
 
-    def evaluate(lookup: Map[String, WdlValue], wdlFunctions: IoFunctionSet) = {
-      node.instantiatedExpression.expression.evaluateValue(lookup, wdlFunctions) flatMap { node.womType.coerceRawValue(_).toErrorOr }
+    def evaluate(lookup: Map[String, WdlValue], ioFunctions: IoFunctionSet) = {
+      instantiatedExpression.expression.evaluateValue(lookup, ioFunctions) flatMap { womType.coerceRawValue(_).toErrorOr }
     }
   }
 
+  case class IntermediateValueKey(node: ExpressionNode, index: ExecutionIndex) extends ExpressionKey {
```

Comment:
I don't know what this is.

Reply:
It's an expression that is not a workflow output. I took the name from what is in wdl4s but I can try to find something more meaningful

Reply:
OK, or maybe just docs if that's easier. 🙂